test/e2e/external_gateways.go

```diff
@@ -0,0 +1,467 @@
+
+		gomega.Eventually(func() bool {
+			_, err := clientSet.CoreV1().Pods(defaultNamespace).Get(context.Background(), gatewayPodName2, metav1.GetOptions{})
+			return apierrors.IsNotFound(err)
+		}, 3*time.Minute, 5*time.Second).Should(gomega.BeTrue(), "Pod was not being deleted")
+	})
+
+	ginkgotable.DescribeTable("Should validate connectivity to an external gateway's loopback address via a pod with external gateway annotations enabled",
+		func(addresses *gatewayTestIPs, icmpCommand string) {
+			if addresses.srcPodIP == "" || addresses.nodeIP == "" {
+				skipper.Skipf("Skipping as pod ip / node ip are not set pod ip %s node ip %s", addresses.srcPodIP, addresses.nodeIP)
+			}
+
+			g := errgroup.Group{}
+			checkPingOnContainer := func(container string) error {
+				_, err := runCommand("docker", "exec", container, "timeout", "60", "tcpdump", "-c", "1", icmpCommand)
+				if err != nil {
+					return err
+				}
+				framework.Logf("ICMP packet successfully detected on gateway %s", container)
+				return nil
+			}
+
+			g.Go(func() error { return checkPingOnContainer(gwContainer1) })
+			g.Go(func() error { return checkPingOnContainer(gwContainer2) })
+
+			// Verify the external gateway loopback address running on the external container is reachable and
+			// that traffic from the source ping pod is proxied through the pod in the default namespace
+			ginkgo.By("Verifying connectivity via the gateway namespace to the remote addresses")
+			for _, t := range addresses.targetIPs {
+				go func(target string) {
+					_, err := framework.RunKubectl(f.Namespace.Name, "exec", srcPingPodName, testContainerFlag, "--", "ping", "-c", testTimeout, target)
+					framework.ExpectNoError(err, "Failed to ping the remote gateway network from pod", target, srcPingPodName)
+				}(t)
+			}
+			err := g.Wait()
+			framework.ExpectNoError(err, "failed to reach the mock gateway(s)")
+		},
+		ginkgotable.Entry("ipv4", &addressesv4, "icmp"),
+		ginkgotable.Entry("ipv6", &addressesv6, "icmp6"))
+
+	ginkgotable.DescribeTable("Should validate connectivity to an external gateway's loopback address via a pod with external gateway annotations enabled",
+		func(protocol string, addresses *gatewayTestIPs, destPort int) {
+			if addresses.srcPodIP == "" || addresses.nodeIP == "" {
+				skipper.Skipf("Skipping as pod ip / node ip are not set pod ip %s node ip %s", addresses.srcPodIP, addresses.nodeIP)
+			}
+
+			expectedHostNames := make(map[string]struct{})
+			for _, c := range []string{gwContainer1, gwContainer2} {
+				res, err := runCommand("docker", "exec", c, "hostname")
+				framework.ExpectNoError(err, "failed to run hostname on", c)
+				hostname := strings.TrimSuffix(res, "\n")
+				expectedHostNames[hostname] = struct{}{}
+			}
+			framework.Logf("Expected hostnames are %v", expectedHostNames)
+
+			returnedHostNames := make(map[string]struct{})
+			target := addresses.targetIPs[0]
+			success := false
+			for i := 0; i < 20; i++ {
```

Comment:
I'm not quite understanding what this loop is doing, I don't see anything incrementing in the loop? What am I missing?

Reply:
What we look for here, is to hit both external containers.
We aim to the same address multiple times, and we collect the hostname of the endpoint we hit. Comparing the returnedHostNames map with the expectedHostNames map assures us that each container was hit at least once.